src/cascadia/TerminalSettingsModel/FileUtils.cpp

```diff
@@ -123,15 +123,41 @@ namespace Microsoft::Terminal::Settings::Model
 
     void WriteUTF8FileAtomic(const std::filesystem::path& path, const std::string_view content)
     {
-        auto tmpPath = path;
+        // GH#10787: rename() will replace symbolic links themselves and not the path they point at.
+        // It's thus important that we first resolve them before generating temporary path.
+        std::error_code ec;
+        auto resolvedPath = std::filesystem::canonical(path, ec);
+        if (ec)
+        {
+            if (ec.value() != ERROR_FILE_NOT_FOUND)
+            {
+                THROW_WIN32_MSG(ec.value(), "failed to compute canonical path");
+            }
+
+            // We were not able to compute the canonical path as the target does not exist.
+            if (std::filesystem::is_symlink(path))
+            {
+                // In the case the original file is a link (but the target doesn't exist), we still need to resolve it.
+                // We considered two fall-backs:
+                // * resolving the link manually (by invoking read_symlink()) which might be less accurate
+                // * writing the file directly, which ensures more accurate link resolution, but leaves the write not atomic.
+                // We have chosen the latter, as this is an edge case and our "atomic" writes are only best-effort anyways.
+                WriteUTF8File(path, content);
+                return;
+            }

```

Comment:
It doesn't need the expensive `canonical()` call if the final component isn't a symlink, which is how you structured it in the first code that I saw. For example:
```c++
    std::error_code ec;
    auto resolvedPath = std::filesystem::is_symlink(path) ?
                          std::filesystem::canonical(path, ec) : path;
    if (ec)
    {
        if (ec.value() != ERROR_FILE_NOT_FOUND)
        {
            THROW_WIN32_MSG(ec.value(), "failed to compute canonical path");
        }
        
        // The original file is a symlink, but the target doesn't exist. 
        // Consider two fall-backs:
        //   * resolve the link manually, which might be less accurate and
        //     more prone to race conditions
        //   * write to the file directly, which lets the system resolve the
        //     the symlink but leaves the write non-atomic
        // The latter is chosen, as this is an edge case and our 'atomic'
        // writes are only best-effort.
        
        WriteUTF8File(path, content);
        return;
    }
```

Reply:
> I was less concerned of the performance of canonical vs is_symlink (both of which require system calls)

That's reasonable. To clarify my "expensive" comment, they both have to open a file, but the open in `is_symlink()` doesn't have to reparse the symlink target.  Once opened it's a single system call to check for a symlink. In contrast, `canonical()` uses the API function `GetFinalPathNameByHandleW()`, which is implemented with multiple system calls. It uses `NtQueryObject()` and `NtQueryInformationFile()` to split the final path into the device part and filesystem part -- e.g. "\\Device\\HarddiskVolume2\\Windows\\explorer.exe" -> ( "\\Device\\HarddiskVolume2",  "\\Windows\\explorer.exe" ). Another `NtQueryInformationFile()` call gets the normalized path in the filesystem -- e.g. get the actual name casing and long filenames for every component in the final path. It opens a handle to "\\\\.\\MountPointManager" via `NtCreateFile()` and queries the mountpoint manager for the canonical DOS name of the opened device via `NtDeviceIoControlFile()` -- e.g. "\\Device\\HarddiskVolume2" -> "\\\\?\\C:".